sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AuthToken.cs

```diff
@@ -0,0 +1,175 @@
+﻿using Microsoft.Rest;
+using System;
+using System.IdentityModel.Tokens.Jwt;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.ContainerRegistry
+{
+    /// <summary>
+    /// Simple authentication class for use with Token related api calls for determining scopes and other such things
+    /// </summary>
+    public class TokenCredentials : ServiceClientCredentials
+    {
+        private string _authHeader { get; set; }
+
+        /*To be used for General Login Scheme*/
+        public TokenCredentials(string username, string password)
+        {
+            _authHeader = EncodeTo64(username + ":" + password);
+        }
+
+        /*To be used for exchanging AAD Tokens for ACR Tokens*/
+        public TokenCredentials()
+        {
+            _authHeader = null;
+        }
+
+        public override async Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException("request");
+            }
+            if (_authHeader != null)
+            {
+                request.Headers.Authorization = new AuthenticationHeaderValue("Basic", _authHeader);
+            }
+            await base.ProcessHttpRequestAsync(request, cancellationToken);
+        }
+
+        static public string EncodeTo64(string toEncode)
+        {
+            byte[] toEncodeAsBytes = System.Text.ASCIIEncoding.ASCII.GetBytes(toEncode);
+            string returnValue = System.Convert.ToBase64String(toEncodeAsBytes);
+            return returnValue;
+        }
+    }
+
+    /// <summary>
+    /// This allows us to have refresh token chains. For example if an access token needs a refresh
+    /// it can use the refresh token which can be refreshed using an aad access token which can be
+    /// refreshed using an aad refresh token which can be obtained using service principals.This
+    /// will all be done internally and thus abstracts much of the checking away.
+    /// </summary>
+    public class AuthToken
+    {
+
+        public delegate string acquireCallback();
+        private static readonly JwtSecurityTokenHandler JwtSecurityClient = new JwtSecurityTokenHandler();
+
+        // Constant to refresh tokens slightly before they are to expire guarding against possible latency related crashes
+        private TimeSpan LATENCY_SAFETY { get; set; } = TimeSpan.FromMinutes(2);
```

Comment:
As this isn't a constant, I would rename this to LatencySafety. 
If this can't be changed, I would also make the field `readonly`